src/workerd/api/worker-rpc.c++

```
@@ -2,19 +2,175 @@
+
+        // Wait for both customEvent and callResult to finish before working with result.
+        auto jsPromise = ioContext.awaitIo(js, kj::joinPromisesFailFast(promises.finish()),
+            [callResult=kj::mv(maybeCallResult), customEventResult=kj::mv(maybeCustomEventResult)]
+            (jsg::Lock& js) -> jsg::Value {
+
+          // Both promises are ready!
+          auto& rpcResult = KJ_REQUIRE_NONNULL(*callResult);
+          auto& customEvent = KJ_REQUIRE_NONNULL(*customEventResult);
+
+          auto serializedResult = rpcResult.getResult().getV8Serialized();
+          auto deserialized = jsg::deserializeV8Rpc(js, kj::heapArray(serializedResult.asBytes()));
+          if (customEvent.outcome == EventOutcome::OK) {
+            // TODO(now): Need to think about how we handle returning vs. throwing error from
+            // remote.
+            if (deserialized.isNativeError()) {
+              JSG_FAIL_REQUIRE(Error, deserialized.toString(js));
+            } else {
+              return jsg::Value(js.v8Isolate, kj::mv(deserialized));
+            }
+          }
+          KJ_UNREACHABLE;
+        });
+        return js.wrapSimplePromise(kj::mv(jsPromise));
+      }
+  ));
+}
+
+kj::Promise<void> JsRpcTargetImpl::call(CallContext callContext) {
+  auto methodName = kj::heapString(callContext.getParams().getMethodName());
+  auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+  // Try to execute the requested method.
+  try {
+    co_await ctx.run(
+        [this,
+         methodName=kj::mv(methodName),
+         serializedArgs = kj::mv(serializedArgs),
+         entrypointName = entrypointName,
+         &callContext] (Worker::Lock& lock) mutable {
+
+      auto reader = callContext.initResults(capnp::MessageSize { 4, 1 }).initResult();
+
+      auto handler = lock.getExportedHandler(entrypointName, ctx.getActor());
+      KJ_IF_SOME(h, handler) {
```

Suggestion: can you fail fast here and throw if we don't have a handler?

---

https://github.com/cloudflare/workerd/compare/e1f09aaa04e65125995e646e2dfc46879e34e5a1..5556c7bce286e193e5956e3e311b439d60956985